### PR TITLE
Verify DVSNI

### PR DESCRIPTION
**Pull request by kuba**
Simple verification for `DVSNI` challenges.

This API is can be used in case of failed challenges (or before sending responses), similarily to what `ManualAuthenticator` does before sending `SimpleHTTP` responses. What do you think, @bradmw?

NB Left `test_verify_cert_postive` unimplemented, because it will have to be changed in #597 anyway... I would like to rebase my current #597 work on top of this PR.

### CODE REVIEW

File: acme/acme/crypto_util.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+import OpenSSL
+
+from acme import errors
+
+
+logger = logging.getLogger(__name__)
+
+
+_DEFAULT_SSL_METHOD = OpenSSL.SSL.SSLv23_METHOD
```
Comment (jdkasten) [inline acme/acme/crypto_util.py]: There are soooo many major problems with SSL.  It should be removed from servers altogether.

We should be using OpenSSL.SSL.TLSv1_METHOD at a minimum.  TLSv1_2 would probably work for our purposes.  We should be dealing with modern TLS clients.

Comment (kuba) [inline acme/acme/crypto_util.py]: I can't think of any SSL vulnerabilities that would affect DVSNI challenge certificate serving / probing. It seems that SNI is all that matters in this challenge and the only attack surface is DOS, since prober needs to check certificate for expected contents anyway; SSL/TLS method used doesn't influence it.

`TLSv1_METHOD` is only compatible with `TLSv1_METHOD`, while `SSLv23_METHOD` is compatible with all other methods, including `TLSv2_METHOD`. https://www.openssl.org/docs/ssl/SSLv23_method.html. Code would have to be rewritten to attempt multiple methods before failing.

If I'm not wrong in the above statement, I would suggest to fix the constant name to `_DEFAULT_DVSNI_SSL_METHOD` and leave the comment inline in the source code.

Comment (bradmw) [inline acme/acme/crypto_util.py]: After looking into this a bit, I think you're right @Kuba. While it probably isn't an issue, perhaps we could use `set_options` to exclude SSLv2 and SSLv3 protocols.

File: acme/acme/crypto_util.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+
+def _serve_sni(certs, sock, reuseaddr=True, method=_DEFAULT_SSL_METHOD,
+               accept=None):
```
Comment (bradmw) [inline acme/acme/crypto_util.py]: This feels out of place to me. This is (basically) what standalone does. If this meant to be an improvement on standalone, why not put the code there in another PR? Why does the acme module need to be able to start a SNI-enabled server?

**Comment by bradmw** [comment]
Sorry for the delay in looking at this @Kuba. I somehow missed this in a flood of e-mails and didn't notice it until yesterday. Thanks for doing this! It'll make testing much easier for me.

File: acme/acme/crypto_util.py (hunk 2 above)
Comment (kuba) [inline acme/acme/crypto_util.py]: 1. `acme` is meant for both client and server implementations
2. This was the only sensible way to unittest `_probe_sni`
3. I'm planning to refactor standalone (including the use of `_serve_sni`) in a separate PR.

**Comment by kuba** [comment]
Thanks for the review. Merged with `master`, renamed and added comment about `_DEFAULT_DVSNI_SSL_METHOD`.

**Comment by kuba** [comment]
Sorry, had to update it with new commits, in particular comment regarding `test_verify_cert_postive` is now obsolote. PTAL @bradmw.

NB https://github.com/kuba/letsencrypt/tree/dvsni-v03 builds on top of this PR

**Comment by bradmw** [comment]
Thanks @kuba! Unless @jdkasten has anything else to add, I'm fine merging this.

**Comment by jdkasten** [comment]
Thanks!
